### ISSUE TITLE
Windows + CMake buildsystem misbehaves when using a mix of internal and external image libraries

### ISSUE THREAD

**Issue by AlexMax** [fltk/fltk#337]
A big shout-out to @Albrecht-S for his prior help.

I have FLTK as a dependency for a game.  The game uses zlib and libPNG, but not libJPEG.  I build FLTK with the exact same settings as my parent project, and additionally...

- `-DOPTION_USE_SYSTEM_LIBJPEG=OFF`
- `-DOPTION_PRINT_SUPPORT=OFF`
- `-DOPTION_USE_GL=OFF`
- `-DFLTK_BUILD_TEST=OFF`
- `-DZLIB_LIBRARY_RELEASE=${CMAKE_CURRENT_BINARY_DIR}/local/lib/${libprefix}zlibstatic${libsuffix}")`
- `-DPNG_PNG_INCLUDE_DIR=${CMAKE_CURRENT_BINARY_DIR}/local/include/libpng16"`
- `-DHAVE_PNG_H=ON`

Two things to note.  First, during the build process, all of my dependencies go into a "local" directory, to keep them together.  Second, the PNG defines are a workaround for FLTK not looking in the correct directory for libpng includes.

This setup works great working on my local Windows 10 machine.  However, as soon as I started pushing my commits to github, the Github Actions starts failing because of this error message:

```
cmd.exe /C "cd . && C:\ProgramData\chocolatey\bin\c++.exe -O2 -g -DNDEBUG -Wl,-subsystem,windows  -mwindows fluid/CMakeFiles/fluid.dir/CodeEditor.cxx.obj fluid/CMakeFiles/fluid.dir/StyleParse.cxx.obj fluid/CMakeFiles/fluid.dir/Fl_Function_Type.cxx.obj fluid/CMakeFiles/fluid.dir/Fl_Group_Type.cxx.obj fluid/CMakeFiles/fluid.dir/Fl_Menu_Type.cxx.obj fluid/CMakeFiles/fluid.dir/Fl_Type.cxx.obj fluid/CMakeFiles/fluid.dir/Fl_Widget_Type.cxx.obj fluid/CMakeFiles/fluid.dir/Fl_Window_Type.cxx.obj fluid/CMakeFiles/fluid.dir/Fluid_Image.cxx.obj fluid/CMakeFiles/fluid.dir/about_panel.cxx.obj fluid/CMakeFiles/fluid.dir/align_widget.cxx.obj fluid/CMakeFiles/fluid.dir/alignment_panel.cxx.obj fluid/CMakeFiles/fluid.dir/code.cxx.obj fluid/CMakeFiles/fluid.dir/factory.cxx.obj fluid/CMakeFiles/fluid.dir/file.cxx.obj fluid/CMakeFiles/fluid.dir/fluid.cxx.obj fluid/CMakeFiles/fluid.dir/function_panel.cxx.obj fluid/CMakeFiles/fluid.dir/pixmaps.cxx.obj fluid/CMakeFiles/fluid.dir/shell_command.cxx.obj fluid/CMakeFiles/fluid.dir/template_panel.cxx.obj fluid/CMakeFiles/fluid.dir/undo.cxx.obj fluid/CMakeFiles/fluid.dir/widget_browser.cxx.obj fluid/CMakeFiles/fluid.dir/widget_panel.cxx.obj fluid/CMakeFiles/fluid.dir/ExternalCodeEditor_WIN32.cxx.obj -o bin\fluid.exe -Wl,--out-implib,bin\libfluid.dll.a -Wl,--major-image-version,0,--minor-image-version,0  lib/libfltk.a  lib/libfltk_images.a  lib/libfltk_forms.a  -lgdiplus  lib/libfltk_jpeg.a  D:/a/odamex/odamex/build-gcc/libraries/local/lib/libzlibstatic.a  C:/Strawberry/c/lib/libpng.a  D:/a/odamex/odamex/build-gcc/libraries/local/lib/libzlibstatic.a  C:/Strawberry/c/lib/libpng.a  lib/libfltk.a  -lcomctl32  -lws2_32  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 && cd ."
lib/libfltk_images.a(Fl_JPEG_Image.cxx.obj): In function `Fl_JPEG_Image::load_jpg_(char const*, char const*, unsigned char const*)':
D:/a/odamex/odamex/libraries/fltk/src/Fl_JPEG_Image.cxx:253: undefined reference to `jpeg_std_error'
D:/a/odamex/odamex/libraries/fltk/src/Fl_JPEG_Image.cxx:273: undefined reference to `jpeg_finish_decompress'
D:/a/odamex/odamex/libraries/fltk/src/Fl_JPEG_Image.cxx:275: undefined reference to `jpeg_destroy_decompress'
D:/a/odamex/odamex/libraries/fltk/src/Fl_JPEG_Image.cxx:298: undefined reference to `jpeg_CreateDecompress'
D:/a/odamex/odamex/libraries/fltk/src/Fl_JPEG_Image.cxx:300: undefined reference to `jpeg_stdio_src'
D:/a/odamex/odamex/libraries/fltk/src/Fl_JPEG_Image.cxx:304: undefined reference to `jpeg_read_header'
D:/a/odamex/odamex/libraries/fltk/src/Fl_JPEG_Image.cxx:311: undefined reference to `jpeg_calc_output_dimensions'
D:/a/odamex/odamex/libraries/fltk/src/Fl_JPEG_Image.cxx:321: undefined reference to `jpeg_start_decompress'
D:/a/odamex/odamex/libraries/fltk/src/Fl_JPEG_Image.cxx:327: undefined reference to `jpeg_read_scanlines'
D:/a/odamex/odamex/libraries/fltk/src/Fl_JPEG_Image.cxx:330: undefined reference to `jpeg_finish_decompress'
D:/a/odamex/odamex/libraries/fltk/src/Fl_JPEG_Image.cxx:331: undefined reference to `jpeg_destroy_decompress'
lib/libfltk_images.a(Fl_JPEG_Image.cxx.obj):Fl_JPEG_Image.cxx:(.rdata$.refptr.jpeg_resync_to_restart[.refptr.jpeg_resync_to_restart]+0x0): undefined reference to `jpeg_resync_to_restart'
collect2.exe: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

This error does not appear to be specific to MinGW, as I get a similar error from the MSVC action as well.

It took me a while to figure out where the issue was coming from, but FLTK apparently prefixes all symbols of its dependencies when using them internally.  Which is cool.  However, this seems like it causes issues if the include directory for another version of the library is found and used.  In this instance, the culprit appears to be something installed on the 2019 version of the Windwos action runner - Strawberry Perl.  This Perl installation is being found and used by CMake as the "system" library, which is not only undesirable, but it introduces the Strawberry Perl include directory into the include stack.

Trying to debug further, I modified my compilation script to dump the CMakeCache of FLTK and then log it where I could see it.  Here it where the Strawberry Perl pollution came in:

```
FREETYPE_PATH:PATH=C:/Strawberry/c/include/freetype2
LIB_freetype:FILEPATH=LIB_freetype-NOTFOUND
PKG_CONFIG_EXECUTABLE:FILEPATH=C:/Strawberry/perl/bin/pkg-config.bat
ZLIB_INCLUDE_DIR:PATH=C:/Strawberry/c/include
FIND_PACKAGE_MESSAGE_DETAILS_PkgConfig:INTERNAL=[C:/Strawberry/perl/bin/pkg-config.bat][v0.26()]
FIND_PACKAGE_MESSAGE_DETAILS_ZLIB:INTERNAL=[D:/a/odamex/odamex/build/libraries/local/lib/zlibstatic.lib][C:/Strawberry/c/include][v1.2.11()]
```

Strangely enough, installing Strawberry Perl locally is not resulting in any problematic behavior on my local system.  However, I am fairly confident that Strawberry Perl is the issue, as updating to the beta Windows 2022 build fixed it, as that buildbot does not have Strawberry Perl installed.

However, I feel like there's a bigger conversation to be had.  On Wndows, I do not think that using a "scanning" version of `find_package` is a wise idea.  Windows does not come with or maintain these libraries, and it is a system built around vendoring your dependencies, not relying on the distro to update them . Thus, to prevent CMake from accidentally picking up on outside libraries unintentionally, I think that the Windows (and possibly Mac) platforms should turn use some of the `NO_PATH` parameters listed in the [CMake docs](https://cmake.org/cmake/help/latest/command/find_package.html#full-signature).  On those platforms, it should either use internal libraries or believe uncritically outside attempts to point at a specific version (like with `CMAKE_PREFIX_PATH`).

For what it's worth, I did manage to pass the following additional parameters to the build to fix the issue specifically for me.

```
-DCMAKE_FIND_USE_PACKAGE_ROOT_PATH=OFF
-DCMAKE_FIND_USE_CMAKE_PATH=OFF
-DCMAKE_FIND_USE_CMAKE_ENVIRONMENT_PATH=OFF
-DCMAKE_FIND_USE_SYSTEM_ENVIRONMENT_PATH=OFF
-DCMAKE_FIND_USE_CMAKE_SYSTEM_PATH=OFF
```

**Comment by Albrecht-S** [comment]
Thank you very much for the report and your help in the long investigation (dialog) on Matrix in the previous days. As we found out linking `fluid` was broken because `src/Fl_JPEG_Image.cxx` was obviously compiled with the "system jpeg" headers and not with FLTK's internal "prefixed" headers.

BTW: it's a **positive effect** of our new *image library prefixing* to see these error messages rather than linking to the wrong (system) library which would have silently ignored a potential version mismatch and caused problems at runtime.

Back to the problem: My current assumption is that this is only a symptom of the underlying problem that our CMake build looks for system jpeg headers although `OPTION_USE_SYSTEM_LIBJPEG=OFF` should disable these. The compile command seemed not to use the necessary "`-I <fltk-root>/jpeg`" commanline switch and the compiler found the system headers as a result.

I need to investigate if my assumption is correct and find the exact details why this happens. If it's true I should be able to fix it in our CMake files.

**Comment by Albrecht-S** [comment]
> However, I feel like there's a bigger conversation to be had. On Wndows, I do not think that using a "scanning" version of find_package is a wise idea. ...

@AlexMax We're open for such discussions in our [developer forum fltk.coredev](https://groups.google.com/g/fltkcoredev) (Google Groups). Please don't follow-up in this thread (issue) because it would distract from the original problem. Thanks

**Comment by AlexMax** [comment]
Just to update the thread, the small throwaway parenthetical I made regarding `CMAKE_PREFIX_PATH` made me double-check my work and I realized that I was not passing that variable to the FLTK build.  Doing that fixed my specific problem, though I'm not sure if you want to keep the issue open or not.

**Comment by Albrecht-S** [comment]
Thanks. And yes, I still want to let the issue open for further investigation.

**Comment by Albrecht-S** [comment]
I believe that this issue has been fixed in commit e80a2734e847bdbbf060b52f86890b409b033710 and I'm closing this issue now.

@AlexMax If you can still test the issue with your "old version" feedback would still be appreciated, whether it still fails or not.
